src/adagrams.js

```diff
@@ -1,15 +1,168 @@
+  }
+  return multiStr;
+};
+//const letterBank = getLetters();
+
 export const drawLetters = () => {
-  // Implement this method for wave 1
+  const randomHand = [];
+  let letterBank = getLetters();
+  for (let i = 0; i < 10; i++) {
+    let randomLetter =
+      letterBank[Math.floor(Math.random() * letterBank.length)];
+    randomHand.push(randomLetter);
+    letterBank = letterBank.replace(randomLetter, "");
+  }
+  return randomHand;
 };
 
 export const usesAvailableLetters = (input, lettersInHand) => {
-  // Implement this method for wave 2
+  let filteredHand = lettersInHand;
+
+  for (let i = 0; i < input.length; i++) {
+    console.log(input[i]);
+
+    if (filteredHand.includes(input[i]) === false) {
+      return false;
+    } else if (filteredHand.includes(input[i]) === true) {
+      let letter = input[i];
+      const index = filteredHand.indexOf(letter);
+      delete filteredHand[index];
+      console.log(filteredHand);
+    }
+  }
+  return true;
 };
 
 export const scoreWord = (word) => {
-  // Implement this method for wave 3
+  let totalScore = 0;
+  let scoredLetter = 0;
+  word = word.toUpperCase();
+
+  for (let i = 0; i < word.length; i++) {
+    scoredLetter = scoreChart[word[i]];
+    totalScore = totalScore + scoredLetter;
+  }
+
+  if (word.length >= 7) {
+    totalScore += 8;
+  }
+  return totalScore;
 };
 
 export const highestScoreFrom = (words) => {
-  // Implement this method for wave 4
+  let scoredWords = {};
+
+  //get item = {scored_word:score}, add item to scoredWords object
+  for (let i = 0; i < words.length; i++) {
+    let scoreCalculated = scoreWord(words[i]);
+    //console.log(scoreCalculated)
+    scoredWords[words[i]] = scoreCalculated;
+    //let scoredWords[words[i]] = scoreCalculated
+  }
+
+  // get max of scores from scoredWords{}
+  let tieList = [];
+  let winningWord = "";
+  let winningScore = 0;
+
+  for (let [word, score] of Object.entries(scoredWords)) {
+    //console.log(scoredWords)
+    if (score > winningScore) {
+      winningScore = score;
+      winningWord = word;
+      console.log(scoredWords);
+      // console.log(winningWord)
+      // console.log(winningScore)
+    }
+    // else if (score === winningScore) {
+    //   tieList.push(winningWord, word);
+    //   console.log(score);
+    // }
+    //ensure that all words in tieList have scores === winningScore
+  }
+
+  //find ties
+  for (let [word, score] of Object.entries(scoredWords)) {
+    if (score === winningScore) {
+      tieList.push(word);
+      console.log(tieList);
+    }
+  }
+  if (tieList.length > 1) {
+    //find min length
+    //return word with min length
+    let sortedList = tieList.sort((a, b) => a.length - b.length);
+    if (sortedList[sortedList.length - 1].length === 10) {
+      winningWord = sortedList[sortedList.length - 1];
+    } else {
+      winningWord = sortedList[0];
+    }
+  }
+
+  const result = { score: winningScore, word: winningWord };
+
+  return result;
 };
```

Comment:
I understand the instinct to build a dictionary with all the scored words and then finding all the ones with the highest score and then making the comparisons separately, but this is one case where breaking everything down into its requisite pieces actually really muddles things and makes the algorithm less efficient. The pseudocode for a better approach is as follows:

     1. Create variables to hold the highest scoring word and the highest score.
     2. Loop through the word list and grab each word and its score.
     3. If its score is higher than the highest score, replace the highest scoring word and highest score.
     4. If there is a tie, do the comparisons for length in place within the loop itself.
     5. Once the loop is finished, you'll have the highest scoring word.